tests/std/tests/P0288R9_move_only_function/test.cpp

```diff
@@ -0,0 +1,393 @@
+            assert(counter::moves == 0);
+        }
+    }
+
+    if constexpr (is_class_v<F>) {
+        assert(counter::inst == 0);
+    }
+}
+
+void test_assign() {
+    pass_this_by_ref x{63};
+
+    {
+        test_function_t f1{small_callable{}};
+        test_function_t f2{large_callable{}};
+        f2 = move(f1);
+        assert(f2(23, x) == 38);
+        f1 = large_callable{};
+        assert(f1(23, x) == 39);
+    }
+
+    {
+        test_function_t f1{large_callable{}};
+        test_function_t f2{small_callable{}};
+        f2 = move(f1);
+        assert(f2(23, x) == 39);
+        f1 = small_callable{};
+        assert(f1(23, x) == 38);
+    }
+
+    {
+        test_function_t f1{small_callable{}};
+        test_function_t f2{odd_cc_callable{}};
+        f2 = move(f1);
+        assert(f2(23, x) == 38);
+        f1 = odd_cc_callable{};
+        assert(f1(23, x) == 40);
+    }
+
+    {
+        test_function_t f1{large_callable{}};
+        test_function_t f2{large_implicit_ptr_callable{}};
+        f2 = move(f1);
+        assert(f2(23, x) == 39);
+        f1 = large_implicit_ptr_callable{};
+        assert(f1(23, x) == 41);
+    }
+
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wself-move"
+#endif // __clang__
+    {
+        test_function_t f1{small_callable{}};
+        test_function_t f2{large_callable{}};
+        f1 = move(f1); // deliberate self-move as a test case
+#pragma warning(suppress : 26800) // use a moved-from object
+        assert(f1(23, x) == 38);
+        f2 = move(f2); // deliberate self-move as a test case
+#pragma warning(suppress : 26800) // use a moved-from object
+        assert(f2(23, x) == 39);
+    }
+#ifdef __clang__
+#pragma clang diagnostic pop
+#endif // __clang__
+}
+
+void test_swap() {
+    pass_this_by_ref x{63};
+
+    {
+        test_function_t f1{small_callable{}};
+        test_function_t f2{large_callable{}};
+        swap(f1, f2);
+        assert(f2(23, x) == 38);
+        assert(f1(23, x) == 39);
+    }
+
+    {
+        test_function_t f1{small_callable{}};
+        test_function_t f2{odd_cc_callable{}};
+        f1.swap(f2);
+        assert(f2(23, x) == 38);
+        assert(f1(23, x) == 40);
+    }
+
+    {
+        test_function_t f1{large_callable{}};
+        test_function_t f2{large_implicit_ptr_callable{}};
+        f2.swap(f1);
+        assert(f2(23, x) == 39);
+        assert(f1(23, x) == 41);
+    }
+
+    {
+        test_function_t f1{small_callable{}};
+        test_function_t f2{large_callable{}};
+        swap(f1, f1);
+        f2.swap(f2);
+        assert(f1(23, x) == 38);
+        assert(f2(23, x) == 39);
+    }
+}
+
+void test_empty() {
+    test_function_t empty;
+    assert(!empty);
+    assert(empty == nullptr);
+    assert(nullptr == empty);
+
+    test_function_t empty_moved = move(empty);
+    assert(!empty_moved);
+    assert(empty_moved == nullptr);
+    assert(!empty_moved);
+    assert(empty_moved == nullptr);

```

Comment:
These lines are repeated. Did you intend to inspect both the source `empty` and the destination `empty_moved`?

Reply:
Yes. Now it looks like too paranoid intention to me (check if moved-out empty function is still empty), but I'll follow it.